File: output/GPT4o-Mini/xCodeEval_4000/C++/Python/fa5bbd62e38df804504d474d91851ba6.py
import sys
import math
from collections import defaultdict

MOD = 998244353

class CPaintTree:
    class Solver:
        def __init__(self, in_stream, out_stream):
            self.in_stream = in_stream
            self.out_stream = out_stream
            self.n = 0
            self.e = []
            self.subtree_sz = []
            self.xy_mapping = []
            self.xys_idx = []
            self.xy = []

        def read_tree(self):
            self.e.clear()
            self.e = [[] for _ in range(self.n)]
            for _ in range(self.n - 1):
                u, v = map(int, self.in_stream.readline().split())
                u -= 1
                v -= 1
                self.e[u].append(v)
                self.e[v].append(u)

        def read_points(self):
            self.xy = [list(map(int, self.in_stream.readline().split())) for _ in range(self.n)]

        def x_dist(self, i, j):
            return self.xy[j][0] - self.xy[i][0]

        def y_dist(self, i, j):
            return self.xy[j][1] - self.xy[i][1]

        def get_angle(self, i, j):
            return math.atan2(self.y_dist(i, j), self.x_dist(i, j))

        def sort_by_angle(self, l, r, cur_xy):
            if r < l:
                return
            for i in range(l, r + 1):
                self.xys_idx[i][0] = self.get_angle(cur_xy, self.xys_idx[i][1])
            max_angle = self.xys_idx[l][0]
            for i in range(l, r + 1):
                max_angle = max(max_angle, self.xys_idx[i][0])
            for i in range(l, r + 1):
                if self.xys_idx[i][0] + math.pi < max_angle:
                    self.xys_idx[i][0] += 2 * math.pi
            self.xys_idx[l:r + 1] = sorted(self.xys_idx[l:r + 1])

        def solve_rec(self, l, r, root, parent):
            curr_xy = self.xys_idx[r][1]
            self.xy_mapping[curr_xy] = root
            self.sort_by_angle(l, r, curr_xy)
            cur_l = l
            for i in self.e[root]:
                if i == parent:
                    continue
                cur_r = cur_l + self.subtree_sz[i] - 1
                nxt_l = cur_r + 1
                self.solve_rec(cur_l, cur_r, i, root)
                cur_l = nxt_l

        def calculate_subtree_sizes(self, i, p):
            self.subtree_sz[i] = 1
            for j in self.e[i]:
                if j == p:
                    continue
                self.calculate_subtree_sizes(j, i)
                self.subtree_sz[i] += self.subtree_sz[j]

        def solve(self):
            self.n = int(self.in_stream.readline().strip())
            self.subtree_sz = [0] * self.n
            self.read_tree()
            self.read_points()
            self.xy_mapping = [0] * self.n
            self.calculate_subtree_sizes(0, -1)
            self.xys_idx = [[0, i] for i in range(self.n)]
            for i in range(self.n):
                self.xys_idx[i][1] = i
            api = lambda i: (self.xy[i][0], self.xy[i][1])
            self.xys_idx.sort(key=lambda i: api(i[1]), reverse=True)
            self.solve_rec(0, self.n - 1, 0, -1)
            self.xy_mapping = [x + 1 for x in self.xy_mapping]
            self.out_stream.write(' '.join(map(str, self.xy_mapping)) + '\n')

    @staticmethod
    def solve(in_stream, out_stream):
        CPaintTree.Solver(in_stream, out_stream).solve()

if __name__ == "__main__":
    CPaintTree.solve(sys.stdin, sys.stdout)

File: output/GPT4o-Mini/xCodeEval_4000/C++/Go/fa5bbd62e38df804504d474d91851ba6.go
package main

import (
	"fmt"
	"math"
	"sort"
)

const MOD = 998244353

type CPaintTree struct{}

type Solver struct {
	n          int
	e          [][]int
	subtreeSz  []int
	xyMapping  []int
	xysIdx     []struct {
		angle float64
		index int
	}
	xy [][]int
}

func (s *Solver) readTree() {
	s.e = make([][]int, s.n)
	for i := 0; i < s.n-1; i++ {
		var u, v int
		fmt.Scan(&u, &v)
		u--
		v--
		s.e[u] = append(s.e[u], v)
		s.e[v] = append(s.e[v], u)
	}
}

func (s *Solver) readPoints() {
	s.xy = make([][]int, s.n)
	for i := 0; i < s.n; i++ {
		s.xy[i] = make([]int, 2)
		fmt.Scan(&s.xy[i][0], &s.xy[i][1])
	}
}

func (s *Solver) xDist(i, j int) int64 {
	return int64(s.xy[j][0]) - int64(s.xy[i][0])
}

func (s *Solver) yDist(i, j int) int64 {
	return int64(s.xy[j][1]) - int64(s.xy[i][1])
}

func (s *Solver) getAngle(i, j int) float64 {
	return math.Atan2(float64(s.yDist(i, j)), float64(s.xDist(i, j)))
}

func (s *Solver) sortByAngle(l, r, curXY int) {
	if r < l {
		return
	}
	for i := l; i <= r; i++ {
		s.xysIdx[i].angle = s.getAngle(curXY, s.xysIdx[i].index)
	}
	maxAngle := s.xysIdx[l].angle
	for i := l; i <= r; i++ {
		if s.xysIdx[i].angle+math.Pi < maxAngle {
			s.xysIdx[i].angle += 2 * math.Pi
		}
	}
	sort.Slice(s.xysIdx[l:r+1], func(i, j int) bool {
		return s.xysIdx[l+i].angle < s.xysIdx[l+j].angle
	})
}

func (s *Solver) solveRec(l, r, root, parent int) {
	currXY := s.xysIdx[r].index
	s.xyMapping[currXY] = root
	s.sortByAngle(l, r, currXY)
	curL := l
	for _, i := range s.e[root] {
		if i == parent {
			continue
		}
		curR := curL + s.subtreeSz[i] - 1
		nxtL := curR + 1
		s.solveRec(curL, curR, i, root)
		curL = nxtL
	}
}

func (s *Solver) calculateSubtreeSizes(i, p int) {
	s.subtreeSz[i] = 1
	for _, j := range s.e[i] {
		if j == p {
			continue
		}
		s.calculateSubtreeSizes(j, i)
		s.subtreeSz[i] += s.subtreeSz[j]
	}
}

func (s *Solver) solve() {
	fmt.Scan(&s.n)
	s.subtreeSz = make([]int, s.n)
	s.readTree()
	s.readPoints()
	s.xyMapping = make([]int, s.n)
	s.calculateSubtreeSizes(0, -1)
	s.xysIdx = make([]struct {
		angle float64
		index int
	}, s.n)
	for i := 0; i < s.n; i++ {
		s.xysIdx[i].index = i
	}
	api := func(i int) (int, int) {
		return s.xy[i][0], s.xy[i][1]
	}
	sort.Slice(s.xysIdx, func(i, j int) bool {
		return api(s.xysIdx[i].index) > api(s.xysIdx[j].index)
	})
	s.solveRec(0, s.n-1, 0, -1)
	for i := 0; i < s.n; i++ {
		s.xyMapping[i]++
	}
	for _, v := range s.xyMapping {
		fmt.Print(v, " ")
	}
	fmt.Println()
}

func (c *CPaintTree) Solve() {
	solver := Solver{}
	solver.solve()
}

func main() {
	var cp CPaintTree
	cp.Solve()
}
